### PR TITLE
[DistGB] save as graphbolt graph directly after partition test case

### PR DESCRIPTION
**Pull request by CfromBU**
modify the test_partiton.py

### CODE REVIEW

**Comment by CfromBU** [comment]
order:
https://github.com/dmlc/dgl/pull/7724 partition test case
https://github.com/dmlc/dgl/pull/7727 partition
https://github.com/dmlc/dgl/pull/7728 dist partition
https://github.com/dmlc/dgl/pull/7690 dist partition test case

File: tests/distributed/test_partition.py [hunk 1]
```diff
@@ -1075,166 +1191,382 @@ def test_not_sorted_node_edge_map():
...
+            # assert F.allclose(
+            #     F.gather_row(g.ndata["h"], local_nodes),
+            #     F.gather_row(part_g.node_attributes["h"], llocal_nodes),
+            # )
+            # assert F.allclose(
+            #     F.gather_row(g.ndata["eh"], local_nodes),
+            #     F.gather_row(part_g.node_attributes["eh"], llocal_nodes),
+            # )
+
+            for name in ["labels", "feats"]:
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:1358]: why not verify via `verify_graph_feats()`?

File: tests/distributed/test_partition.py [hunk 2]
```diff
@@ -1075,166 +1191,382 @@ def test_not_sorted_node_edge_map():
...
+
+                for etype in hg.canonical_etypes:
+                    name = _etype_tuple_to_str(etype) + "/trainer_id"
+                    assert name in edge_feats
+                    part_ids = F.floor_div(
+                        edge_feats[name], num_trainers_per_machine
+                    )
+                    assert np.all(F.asnumpy(part_ids) == part_id)
+
+            # Verify the mapping between the reshuffled IDs and the original IDs.
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:1479]: is such check covered in `verify_hetero_graph()`?

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]

File: python/dgl/distributed/partition.py [hunk 3]
```diff
@@ -109,19 +109,40 @@ def _save_graphs(filename, g_list, formats=None, sort_etypes=False):
...
-        dtype = F.dtype(graph.ndata["inner_node"])
-        return (
-            graph.ndata["inner_node"]
-            * F.astype(graph.ndata[NTYPE] == ntype_id, dtype)
-            == 1
-        )
+def _get_inner_node_mask(graph, ntype_id, use_graphbolt=False):
+    if use_graphbolt:
+        if NTYPE in graph.node_attributes:
+            dtype = F.dtype(graph.node_attributes["inner_node"])
```
[the remark this answers is not in the document]
Comment (CfromBU) [reply python/dgl/distributed/partition.py:115]: before `_get_inner_node_mask` is called, `graph` will be checked if "inner_node" exists. And I added an assert in `_get_inner_node_mask`

File: python/dgl/distributed/partition.py [hunk 4]
```diff
@@ -109,19 +109,40 @@ def _save_graphs(filename, g_list, formats=None, sort_etypes=False):
...
+            )
+        else:
+            return graph.ndata["inner_node"] == 1
 
 
-def _get_inner_edge_mask(graph, etype_id):
+def _get_inner_edge_mask(graph, etype_id, use_graphbolt=False):
+    if use_graphbolt:
+        if graph.type_per_edge is not None:
+            dtype = F.dtype(graph.edge_attributes["inner_edge"])
```
[the remark this answers is not in the document]
Comment (CfromBU) [reply python/dgl/distributed/partition.py:138]: I have remove the ditto.

File: tests/distributed/test_partition.py [hunk 5]
```diff
@@ -35,12 +37,19 @@
...
 
 
-def _verify_partition_data_types(part_g):
-    for k, dtype in RESERVED_FIELD_DTYPE.items():
-        if k in part_g.ndata:
-            assert part_g.ndata[k].dtype == dtype
-        if k in part_g.edata:
-            assert part_g.edata[k].dtype == dtype
+def _verify_partition_data_types(part_g, use_graphbolt=False):
+    if not use_graphbolt:
```
[the remark this answers is not in the document]
Comment (CfromBU) [reply tests/distributed/test_partition.py:41]: Ok, I have replace the duplicate codes.

File: tests/distributed/test_partition.py [hunk 6]
```diff
@@ -81,11 +90,58 @@ def create_random_hetero():
...
+                )
+            )
+            assert g.num_edges(etype) == num_edges[etype]
+
+        nids = {ntype: [] for ntype in g.ntypes}
+        eids = {etype: [] for etype in g.canonical_etypes}
+        for part in parts:
+            eid = th.arange(len(part.edge_attributes[dgl.EID]))
+            etype_arr = F.gather_row(part.type_per_edge, eid)
+            eid_type = F.gather_row(part.edge_attributes[dgl.EID], eid)
```
[the remark this answers is not in the document]
Comment (CfromBU) [reply tests/distributed/test_partition.py:120]: `eid_type` is used to build `eids`, which stores `eids` according to its etype. But `eid_type` is used to connect to the `eids` `etype_arr`'s order. I copied it from source code but the variable `eid_type` doesn't have an appropriate name, so I rename it to `eid_arr` to emphasize its connection with `eid_type`.

**Comment by CfromBU** [comment]
I modified the function in test case that relevant to graphbolt, but didn't change the function about DGLGraph.

File: python/dgl/distributed/partition.py [hunk 7]
```diff
@@ -106,31 +106,40 @@ def _save_graphs(filename, g_list, formats=None, sort_etypes=False):
...
+    )
     save_graphs(filename, g_list, formats=formats)
 
 
-def _get_inner_node_mask(graph, ntype_id):
-    if NTYPE in graph.ndata:
-        dtype = F.dtype(graph.ndata["inner_node"])
+def _get_inner_node_mask(graph, ntype_id, use_graphbolt=False):
+    ndata = graph.node_attributes if use_graphbolt else graph.ndata
+    assert "inner_node" in ndata, '"inner_node" is not nodes\' data'
```
Comment (Rhett-Ying) [inline python/dgl/distributed/partition.py:121]: use `""` instead `'` for the out-most wrapper.

File: python/dgl/distributed/partition.py [hunk 8]
```diff
@@ -106,31 +106,40 @@ def _save_graphs(filename, g_list, formats=None, sort_etypes=False):
...
-        )
+def _get_inner_edge_mask(graph, etype_id, use_graphbolt=False):
+    edata = graph.edge_attributes if use_graphbolt else graph.edata
+    etype = (
+        graph.type_per_edge
+        if use_graphbolt
+        else (graph.edata[ETYPE] if ETYPE in graph.edata else None)
+    )
+    if etype is not None:
+        dtype = F.dtype(edata["inner_edge"])
```
Comment (Rhett-Ying) [inline python/dgl/distributed/partition.py:139]: assert `inner_edge`?

File: python/dgl/distributed/partition.py [hunk 9]
```diff
@@ -444,6 +454,105 @@ def load_partition_feats(
     return node_feats, edge_feats
 
 
+def _load_partition_book_from_metadata(part_metadata, part_id):
+    assert "num_parts" in part_metadata, "num_parts does not exist."
```
Comment (Rhett-Ying) [inline python/dgl/distributed/partition.py:458]: this huge code block is duplicated?

File: tests/distributed/test_partition.py [hunk 10]
```diff
@@ -5,10 +5,12 @@
 import dgl
 
 import dgl.backend as F
+import dgl.sparse as dglsp
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:8]: is it possible to avoid import dgl sparse?

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]

File: tests/distributed/test_partition.py [hunk 11]
```diff
@@ -1160,81 +1665,111 @@ def test_partition_graph_graphbolt_hetero(
...
+            num_trainers_per_machine=1,
             use_graphbolt=True,
             store_eids=store_eids,
             store_inner_node=store_inner_node,
             store_inner_edge=store_inner_edge,
             n_jobs=n_jobs,
         )
+        _vertify_original_IDs(hg, orig_nids, orig_eids)
+
+        if debug_mode:
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:1720]: ditto

File: tests/distributed/test_partition.py [hunk 12]
```diff
@@ -1160,81 +1665,111 @@ def test_partition_graph_graphbolt_hetero(
...
-            if debug_mode:
-                assert th.equal(
-                    orig_g.edata[dgl.ETYPE][orig_eids],
-                    new_g.edge_attributes[dgl.ETYPE],
-                )
-            else:
-                assert dgl.ETYPE not in new_g.edge_attributes
-            assert th.equal(
-                orig_g.edata[dgl.ETYPE][orig_eids], new_g.type_per_edge
+            # TODO verify mapping IDs
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:1734]: TODO?

**Comment by Rhett-Ying** [comment]
@CfromBU pls fix the lint check failure

File: python/dgl/distributed/partition.py [hunk 13]
```diff
@@ -109,28 +109,38 @@ def _save_graphs(filename, g_list, formats=None, sort_etypes=False):
...
 
-def _get_inner_node_mask(graph, ntype_id):
-    if NTYPE in graph.ndata:
-        dtype = F.dtype(graph.ndata["inner_node"])
-        return (
-            graph.ndata["inner_node"]
-            * F.astype(graph.ndata[NTYPE] == ntype_id, dtype)
-            == 1
+def _get_inner_node_mask(graph, ntype_id, use_graphbolt=False, gpb=None):
+    ndata = graph.node_attributes if use_graphbolt else graph.ndata
```
Comment (Rhett-Ying) [inline python/dgl/distributed/partition.py:113]: we could check `isinstance(graph, gb.FusedCSCSamplingGraph)`, then we don't need `use_graphbolt`

File: tests/distributed/test_partition.py [hunk 14]
```diff
@@ -35,12 +35,19 @@
 from utils import reset_envs
 
 
-def _verify_partition_data_types(part_g):
+def _verify_partition_data_types(part_g, use_graphbolt=False):
+    """
+    check list:
+        make sure nodes and edges have correct type.
+    """
+    ndata = part_g.node_attributes if use_graphbolt else part_g.ndata
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:43]: ditto, check isinstance()?

File: tests/distributed/test_partition.py [hunk 15]
```diff
@@ -81,98 +88,220 @@ def create_random_hetero():
...
+    g, part, eids, use_graphbolt, store_inner_edge
+):
+    edata = part.edge_attributes if use_graphbolt else part.edata
+    etype = part.type_per_edge if use_graphbolt else edata[dgl.ETYPE]
+    eid = th.arange(len(edata[dgl.EID]))
+    etype_arr = F.gather_row(etype, eid)
+    eid_arr = F.gather_row(edata[dgl.EID], eid)
+    for etype in g.canonical_etypes:
+        etype_id = g.get_etype_id(etype)
+        eids[etype].append(F.boolean_mask(eid_arr, etype_arr == etype_id))
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:170]: where is `eids` defined? and used? seems not used at all?

File: tests/distributed/test_partition.py [hunk 16]
```diff
@@ -81,98 +88,220 @@ def create_random_hetero():
...
-def verify_graph_feats(
-    g, gpb, part, node_feats, edge_feats, orig_nids, orig_eids
+        )
+        nids[ntype].append(inner_nids)
+
+
+def _verify_node_edge_included_hetero(
+    g, nids, eids, store_inner_edge, store_inner_node
+):
+    if store_inner_node:
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:213]: this body looks weird. How is the `store_inner_node/edge` flag related to the check here? In intuition, this flag is about `partition.ndadta['inner_node/edge']. but it's not directly used here.

File: tests/distributed/test_partition.py [hunk 17]
```diff
@@ -789,26 +1002,20 @@ def test_dgl_partition_to_graphbolt_homo(
...
-            else:
-                assert "inner_node" not in new_g.node_attributes
             if store_eids or debug_mode:
                 assert orig_g.edata[dgl.EID].dtype == th.int64
                 assert new_g.edge_attributes[dgl.EID].dtype == th.int64
                 assert th.equal(
                     orig_g.edata[dgl.EID][orig_eids],
                     new_g.edge_attributes[dgl.EID],
                 )
-            else:
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:801]: why remove these lines? do they fail the testcase? if not , keep it.

File: tests/distributed/test_partition.py [hunk 18]
```diff
@@ -1103,46 +1519,305 @@ def test_partition_graph_graphbolt_homo(
...
+            graph_name,
+            num_parts,
+            store_inner_node,
+            store_inner_edge,
+            store_eids,
+            is_homo=True,
+        )
+
+
+def _verify_builded_id_gb(part_sizes, gpb):
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:1542]: `built`? or `constructed`?

File: tests/distributed/test_partition.py [hunk 19]
```diff
@@ -81,98 +97,220 @@ def create_random_hetero():
...
+        # Make sure edge Ids fall into a range.
+        inner_edge_mask = _get_inner_edge_mask(part, etype_id)
+        inner_eids = np.sort(
+            F.asnumpy(F.boolean_mask(edata[dgl.EID], inner_edge_mask))
+        )
+        assert np.all(
+            inner_eids == np.arange(inner_eids[0], inner_eids[-1] + 1)
         )
-        assert g.num_edges(etype) == num_edges[etype]
+    return eids
```
Comment (Rhett-Ying) [inline tests/distributed/test_partition.py:202]: It's weird that a verify function returns data. As it's tests, let's tolerate it.

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**